rlpx.md

```diff
@@ -42,118 +60,305 @@ the authenticity of the message by checking whether
-stream becomes straight-forward. Frames are authenticated via key material which is
-generated during the handshake.
+## Message Encoding
 
-The frame header provides information about the size of the packet and the packet's source
-protocol.
+Messages are encoded as follows:
 
 ```text
-frame = header || header-mac || frame-data || frame-mac
-header = frame-size || header-data || padding
-frame-size = size of frame excluding padding, integer < 2**24, big endian
-header-data = rlp.list(protocol-type[, context-id])
-protocol-type = integer < 2**16, big endian
-context-id = integer < 2**16, big endian
-padding = zero-fill to 16-byte boundary
-frame-content = any binary data
+frame-data = msg-id || msg-data
+frame-size = length of frame-data, encoded as a 24bit big-endian integer
+```
 
-header-mac = left16(egress-mac.update(aes(mac-secret,egress-mac)) ^ header-ciphertext).digest
-frame-mac = left16(egress-mac.update(aes(mac-secret,egress-mac)) ^ left16(egress-mac.update(frame-ciphertext).digest))
-egress-mac = keccak256 state, continuously updated with egress bytes
-ingress-mac = keccak256 state, continuously updated with ingress bytes
+where `msg-id` is an RLP-encoded integer identifying the message and `msg-data` is an RLP
+list containing the message data.
 
-left16(x) is the first 16 bytes of x
-|| is concatenate
-^ is xor
+### Snappy Compression
+
+All messages following [Hello] are compressed using the Snappy algorithm. Note that the
+`frame-size` of compressed messages refers to the uncompressed size of `msg-`. The
+compressed encoding of messages is as follows:
+
+```text
+frame-data = msg-id || snappyCompress(msg-data)
+frame-size = length of (msg-id || msg-data) encoded as a 24bit big-endian integer
 ```
 
-Message authentication is achieved by continuously updating `egress-mac` or `ingress-mac`
-with the ciphertext of bytes sent (egress) or received (ingress); for headers the update
-is performed by xoring the header with the encrypted output of it's corresponding mac (see
-header-mac above for example). This is done to ensure uniform operations are performed for
-both plaintext mac and ciphertext. All macs are sent cleartext.
+### Message ID-based Multiplexing
+
+While the framing layer supports a `capability-id`, the current version of RLPx doesn't
+use that field for multiplexing between different capabilities. Instead, multiplexing
+relies purely on the message ID.
+
+Each capability is given as much of the message-ID space as it needs. All such
+capabilities must statically specify how many message IDs they require. On connection and
+reception of the [Hello] message, both peers have equivalent information about what
+capabilities they share (including versions) and are able to form consensus over the
+composition of message ID space.
+
+Message IDs are assumed to be compact from ID 0x10 onwards (0x00-0x10 is reserved for the
+"p2p" capability) and given to each shared (equal-version, equal name) capability in
```

Comment:
Capability comparison is case sensitive (unfortunately). If I recall correctly, current implementations see LES and les as different protocols. Something in the spec should note what the situation currently is and also what it should be.